### PR TITLE
Implement the new (and final) display API.

### PR DESCRIPTION
**Pull request by markshannon**
In summary the new API is 
```python
display.show(image, delay=0, wait=True, loop=False, clear=False)
display.show(iterable, delay=400, wait=True, loop=False, clear=False)
display.scroll(string, delay=400)
```
See for https://github.com/bbcmicrobit/micropython/issues/104 for full discussion about the new API.

### CODE REVIEW

**Comment by dpgeorge** [comment]
It looks like display.animate still exists, is that correct?  If it were removed could one also remove the microbit_display_animate helper function, or is that needed by other functions?

**Comment by ntoll** [comment]
Just to help me plan my work over the coming days... is there an ETA when this will land? I'll need to find time to revise the documentation and downstream editors that rely on there being a recent .hex file to flash the device. The release of the actual device is (at a guess) likely to be a matter of a few weeks at the most.

**Comment by markshannon** [comment]
Yes, `display.animate` still exists. 
It could be removed, I don't care much either way. 
What do you think? 
@ntoll?

It won't allow us to remove the `microbit_display_animate` function as that sets up animations in general, but we will able to remove the image slice iterator as this is the only use of it.

**Comment by markshannon** [comment]
I'll have time to fix up any comments tomorrow evening. I'd guess it should be in by the weekend?

**Comment by ntoll** [comment]
@markshannon for `display.animate` we should follow the Zen of Python so that there is one obvious way to do a thing. Ergo please remove the `animate` function. Also, perfect timing so I can do a documentation blitz over the weekend. :-)

**Comment by dpgeorge** [comment]
> we should follow the Zen of Python so that there is one obvious way to do a thing. Ergo please remove the animate function

+1.  Also remove the image slice stuff.

**Comment by markshannon** [comment]
Done. And I've updated the docs as well.

**Comment by ntoll** [comment]
@markshannon care to squash/rebase..? I'll test it works but leave it to @dpgeorge to pass judgement on the C (and hopefully merge)... ;-)

**Comment by markshannon** [comment]
Rebased.

**Comment by ntoll** [comment]
OK... this was also a lot of fun to test and it "feels" nice and simple. Really great work @markshannon, top notch stuff. :-) @dpgeorge waiting on you to check the C and merge. :-D

**Comment by dpgeorge** [comment]
I tested it and it's really nice: simple yet powerful, and nice and consistent.  I can confirm that it fixes #157.

I'll make a few comments in-line.  Then if @markshannon can please rebase I think it's good to merge!

File: source/lib/utils/iters.c [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+} repeat_iterator_t;
+
+STATIC mp_obj_t microbit_repeat_iter_next(mp_obj_t iter_in) {
+    repeat_iterator_t *iter = (repeat_iterator_t *)iter_in;
+    mp_obj_t result = mp_iternext(iter->iterator);
+    if (result == MP_OBJ_STOP_ITERATION) {
+        iter->iterator = mp_getiter(iter->iterable);
+        result = mp_iternext(iter->iterator);
+        if (result == MP_OBJ_STOP_ITERATION) {
+            nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "cannot repeat empty sequence."));
```
Comment (dpgeorge) [inline source/lib/utils/iters.c]: No need for fullstop at end of message.

File: source/microbit/microbitimage.cpp [hunk 2]
```diff
@@ -440,16 +440,6 @@ mp_obj_t microbit_image_invert(mp_obj_t self_in) {
 }
 MP_DEFINE_CONST_FUN_OBJ_1(microbit_image_invert_obj, microbit_image_invert);
 
-mp_obj_t microbit_image_slice_func(mp_uint_t n_args, const mp_obj_t *args) {
```
Comment (dpgeorge) [inline source/microbit/microbitimage.cpp]: There is still a lot of other image slice functionality that can be deleted (eg microbit_image_slice function).

**Comment by dpgeorge** [comment]
The inc/lib/utils and source/lib/utils directory is there to mirror upstream uPy, and has source code from there.  I think we should keep our custom, generic libraries just in inc/lib and source/lib.

**Comment by markshannon** [comment]
All done, and rebased.

**Comment by ntoll** [comment]
Merged :tada:

**Pull request merged**